impress-org/givewp
# Fix: improve PayPal Donation token handling and refreshing

## Description


I implement logic to validate the access token when requesting a new access token to PayPal comments. A log entry will generate on failure. The token refresh action will run every 5 minutes until the website get a valid PayPal Commerce access token.

Paypal HTTP client from Paypal SDK forces fresh access tokens on each donation request. I implemented logic to use stored access tokens.

## Affects


This pull request will affect the following:
- PayPal commerce onboarding
- Refresh token access cron job.
- Ontime donation and subscription processing.

## Testing Instructions


- User should be able to onboard.
- Access token should refresh before half an hour of expiration time.
- Donors should be able to create subscriptions.
- Donors should able to donate one-time donations with credit on the PayPal Express account connected website. 

## Additional Changes
I discovered that a seller access token must connect the user to the same PayPal Rest APP. For this reason, we will keep it for future use. The seller access token does not update, but in the future, we can use a refresh token to update the seller access token.

## Pre-review Checklist






---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204019794141811
